### PR TITLE
resolve: fixes for sd-resolved bypass

### PR DESCRIPTION
**Pull request by rpigott**
Enabled validation for bypass transactions and don't serve cached cd bits.

I might be missing some historical context, but bypass is also enabled for DO packets:

https://github.com/systemd/systemd/blob/a417592b068eec06ccef55c74622e49ae69a762c/src/resolve/resolved-dns-transaction.c#L843-L846

Still, the cd bit in this case was set by the stub during a previous, validated, transaction. So in each case the erroneous reply is still
known to be authentic..

@poettering please review.

Fixes: #34649

### CODE REVIEW

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [inline src/resolve/resolved-dns-stub.c]: i'd write this as:

```c
DNS_PACKET_HEADER(c)->flags = htobe16((be16toh(DNS_PACKET_HEADER(c)->flags) & ~DNS_PACKET_FLAG_AD) | (authenticated ? DNS_PACKET_FLAG_AD : 0));
```

or so

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [inline src/resolve/resolved-dns-stub.c:985]: hmm, is this right? shouldn't this be dependent on the CD flag in the query?

**Comment by poettering** [comment]
> I might be missing some historical context, but bypass is also enabled for DO packets:

yeah, if clients want DNSSEC RRs we better step out of the way and give them exactly what upstream DNS servers gave us.

**Comment by poettering** [comment]
any reason you marked this as draft btw? (i usually don't look at draft stuff with the same priority as on other stuff)

**Comment by rpigott** [comment]
> any reason you marked this as draft btw?

At least the quoted comment is clearly incorrect but I didn't change it.

I guess I'm not really 100% clear on the purpose or use of bypass mode, so I wanted to make sure this wasn't merged before you got a chance to see it, like #34037.

File: src/resolve/resolved-dns-transaction.c
Comment (yuwata) [inline src/resolve/resolved-dns-transaction.c]: Spurious line wrapping...
The line is not so long, please drop the line break.
Or, at least, please wrap the line after `&&`.

File: src/resolve/resolved-dns-stub.c
Comment (yuwata) [inline src/resolve/resolved-dns-stub.c]: Uhh, not a big matter, but also line break is spurious. How about
```
DNS_PACKET_HEADER(c)->flags = htobe16(
                be16toh((DNS_PACKET_HEADER(c)->flags) & ~DNS_PACKET_FLAG_AD) |
                (authenticated ? DNS_PACKET_FLAG_AD : 0));
```
or
```
DNS_PACKET_HEADER(c)->flags = htobe16(be16toh((DNS_PACKET_HEADER(c)->flags) & ~DNS_PACKET_FLAG_AD) |
                                      (authenticated ? DNS_PACKET_FLAG_AD : 0));
```
?

**Comment by yuwata** [comment]
Is it possible to add a test case for this ?

File: src/resolve/resolved-dns-stub.c
Comment (rpigott) [reply src/resolve/resolved-dns-stub.c]: ok #1 looks good

**Comment by yuwata** [comment]
CIs are red:
```
TEST-75-RESOLVED.sh[724]: + run dig +nostats +cd signed.test
TEST-75-RESOLVED.sh[798]: + dig +nostats +cd signed.test
TEST-75-RESOLVED.sh[799]: + tee /tmp/tmp.iZwtvjLbXg
TEST-75-RESOLVED.sh[799]: ; <<>> DiG 9.20.2 <<>> +nostats +cd signed.test
TEST-75-RESOLVED.sh[799]: ;; global options: +cmd
TEST-75-RESOLVED.sh[799]: ;; Got answer:
TEST-75-RESOLVED.sh[799]: ;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 57883
TEST-75-RESOLVED.sh[799]: ;; flags: qr rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
TEST-75-RESOLVED.sh[799]: ;; OPT PSEUDOSECTION:
TEST-75-RESOLVED.sh[799]: ; EDNS: version: 0, flags:; udp: 65494
TEST-75-RESOLVED.sh[799]: ;; QUESTION SECTION:
TEST-75-RESOLVED.sh[799]: ;signed.test.			IN	A
TEST-75-RESOLVED.sh[799]: ;; ANSWER SECTION:
TEST-75-RESOLVED.sh[799]: signed.test.		7199	IN	A	10.0.0.10
TEST-75-RESOLVED.sh[724]: + grep -qF 10.0.0.10 /tmp/tmp.iZwtvjLbXg
TEST-75-RESOLVED.sh[724]: + grep -q 'flags:[^;]* cd'
```

**Review by yuwata — APPROVED** [review]
Looks superficially good, but I am not familiar with the code.
Let's wait for @poettering's final review.

**Comment by poettering** [comment]
so, hmm, taking a step back on this i *do* wonder what the right approach here is. the idea of the bypass stuff originally was that we bypass all our processing, and just forward packets, if client's make clear they want to do their own authentication anyway, via announcing dnssec support. this was then modified so that we suppress the AD bit while forwarding, which made some sense I guess. with this now we start setting the AD bit too while forwarding, which then does raise the question how much of a "bypass" this all still is...

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [inline src/resolve/resolved-dns-stub.c]: maybe:

```c
DNS_PACKET_HEADER(c)->flags = htobe16(UPDATE_FLAG(be16toh(DNS_PACKET_HEADER(c)->flags), DNS_PACKET_FLAG_CD, checking_disabled);
```

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [inline src/resolve/resolved-dns-stub.c]: ditto

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [inline src/resolve/resolved-dns-stub.c]: can you maybe prefix this with `/* checking_disabled= */` to make clear what this is

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [inline src/resolve/resolved-dns-stub.c:468]: why this change?

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [inline src/resolve/resolved-dns-stub.c:734]: both these flags should only be set if dnssec is on, no? for compat reasons?

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [reply src/resolve/resolved-dns-stub.c]: we set both CD and AD only if DO is on in the question, no?

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [inline src/resolve/resolved-dns-stub.c:996]: this should not be necessary. if client sets DO we automatically set bypass mode, i.e. get the upper branch. And if it turns off DO we should not set CD nor AD I guess

File: src/resolve/resolved-dns-transaction.c
Comment (poettering) [inline src/resolve/resolved-dns-transaction.c:844]: hmm, hmm what does this have to do with EDE? can you explain?

Comment (rpigott) [reply src/resolve/resolved-dns-transaction.c:844]: There were some tests failing before I added this that tested usage of the cached packet to forward EDE errors. Previously with dnssec enabled most requests from the stub have +cd so most packets were cached and this test did not fail, but with the updated condition (DO only) it did, so I had to cache those packets with EDE set as well.

File: src/resolve/resolved-dns-stub.c
Comment (rpigott) [reply src/resolve/resolved-dns-stub.c:996]: Why not? I think a client could support these bits without supporting DO.

File: src/resolve/resolved-dns-stub.c
Comment (rpigott) [reply src/resolve/resolved-dns-stub.c:468]: Why transform the packet if we don't have to? I don't think there is any compat concern here, and with this patch we will now update the +cd bit to the "right" value anyway a few lines down, so I don't think this is necessary.

File: src/resolve/resolved-dns-stub.c
Comment (rpigott) [reply src/resolve/resolved-dns-stub.c]: On second thought, I think it's fine to have the reply with +cd and +ad in the case the client request is +cd and we happen to return authenticated data from the cache. We won't be doing that because of the SD_RESOLVED_NO_CACHE being set for +cd packets, but still I don't think it's necessary to check both conditions here, so I'll update it to just the `authenticated` value.

**Comment by poettering** [comment]
> so, hmm, taking a step back on this i _do_ wonder what the right approach here is. the idea of the bypass stuff originally was that we bypass all our processing, and just forward packets, if client's make clear they want to do their own authentication anyway, via announcing dnssec support. this was then modified so that we suppress the AD bit while forwarding, which made some sense I guess. with this now we start setting the AD bit too while forwarding, which then does raise the question how much of a "bypass" this all still is...

but hmm, maybe i am actually fine if we mangle the AD bit to match our authentication results as long as we pass the rest through as is

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [reply src/resolve/resolved-dns-stub.c:468]: where do we do that? dns_stub_finish_reply_packet() is not used in the bypass codepath? or am i blind?

File: src/resolve/resolved-dns-stub.c
Comment (poettering) [reply src/resolve/resolved-dns-stub.c:996]: well, they aren't defined in dnssec-less DNS...

Comment (poettering) [reply src/resolve/resolved-dns-stub.c:996]: just to not risk incompatibilities i'd not set flags that the non-dnssec RFCs didn't know in non-dnssec mode.

File: src/resolve/resolved-dns-stub.c
Comment (rpigott) [reply src/resolve/resolved-dns-stub.c:468]: Uh you're right. I forgot the context of this change and thought I was looking at a section of patch_bypass in the GH review :|.

I still think the cd bit could be set (if our validation is also disabled), but I will remove this change from the patch.

File: src/resolve/resolved-dns-stub.c
Comment (rpigott) [reply src/resolve/resolved-dns-stub.c:996]: Hm... But in this case the client has already set CD, I'm just adjusting the SD_RESOLVED_* flags to implement that request. So clearly the client supports CD here, even if it has not set DO, right? If you're right and no client ever sets CD without also setting DO I guess this line does nothing, but that doesn't seem like a disaster.

File: src/resolve/resolved-dns-stub.c
Comment (rpigott) [reply src/resolve/resolved-dns-stub.c:468]: I've changed my mind again, I think this change does belong. I think a client that does `dig @_localdnsstub +cd example.com` should get a reply with the cd bit set if validation was disabled. I don't know of any RFC recommendation one way or another, but at least the 3 major public resolvers are happy to reply to me with CD and no DO bit set.

Comment (poettering) [reply src/resolve/resolved-dns-stub.c:468]: please split this out into a separate commit, it has nothing to do with the proxy

**Pull request merged**
